src/librbd/image/CreateRequest.cc

```diff
@@ -0,0 +1,696 @@
+  // deadlocks).
+
+  uint64_t snap_id;
+  int r = m_ioctx.selfmanaged_snap_create(&snap_id);
+  if (r == -EINVAL) {
+    lderr(m_cct) << "pool not configured for self-managed RBD snapshot support"
+                 << dendl;
+    *result = r;
+    return m_on_finish;
+  } else if (r < 0) {
+    lderr(m_cct) << "failed to allocate self-managed snapshot: "
+                 << cpp_strerror(r) << dendl;
+    *result = r;
+    return m_on_finish;
+  }
+
+  r = m_ioctx.selfmanaged_snap_remove(snap_id);
+  if (r < 0) {
+    lderr(m_cct) << "failed to release self-managed snapshot " << snap_id
+                 << ": " << cpp_strerror(r) << dendl;
+    *result = r;
+    return m_on_finish;
+  }
+
+  create_id_object();
+  return nullptr;
+}
+
+template<typename I>
+void CreateRequest<I>::create_id_object() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectWriteOperation op;
+  op.create(true);
+  cls_client::set_id(&op, m_imageid);
+
+  using klass = CreateRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_create_id_object>(this);
+  int r = m_ioctx.aio_operate(m_id_obj, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+Context *CreateRequest<I>::handle_create_id_object(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error creating RBD id object: " << cpp_strerror(*result) << dendl;
+    return m_on_finish;
+  }
+
+  add_image_to_directory();
+  return nullptr;
+}
+
+template<typename I>
+void CreateRequest<I>::add_image_to_directory() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectWriteOperation op;
+  cls_client::dir_add_image(&op, m_imgname, m_imageid);
+
+  using klass = CreateRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_add_image_to_directory>(this);
+  int r = m_ioctx.aio_operate(RBD_DIRECTORY, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+Context *CreateRequest<I>::handle_add_image_to_directory(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error adding image to directory: " << cpp_strerror(*result) << dendl;
+
+    m_r_saved = *result;
+    remove_id_object();
+    return nullptr;
+  }
+
+  create_image();
+  return nullptr;
+}
+
+template<typename I>
+void CreateRequest<I>::create_image() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  ostringstream oss;
+  oss << RBD_DATA_PREFIX << m_imageid;
+
+  librados::ObjectWriteOperation op;
+  op.create(true);
+  cls_client::create_image(&op, m_size, m_order, m_features, oss.str());
+  if ((m_stripe_unit || m_stripe_count) &&
+      ((m_stripe_count != 1) || (m_stripe_unit != (1ull << m_order)))) {
+    cls_client::set_stripe_unit_count(&op, m_stripe_unit, m_stripe_count);
```

Comment:
Well, I would have been happy if it returned `-ENOENT` -- what we actually get is `-ENOEXEC` from `require_feature()` in `set_stripe_unit_count()` which happens *after* a `stat()` check on the object. So, the (header) object exist, but reading omap value (`features` key in this case) fails.